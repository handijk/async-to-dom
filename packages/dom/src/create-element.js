import { SPECIAL_PLACEHOLDER_ITEM } from './constants.js';
import { ELEMENT_RENDERERS } from './element-renderers/get-default-renderers.js';
import { ATTRIBUTE_RENDERERS } from './attribute-renderers/get-attribute-renderers.js';
import { render } from '@async-to-html/render/render.js';

export const createElement =
  ({ document } = {}) =>
  (tag, attributes = {}, ...children) => ({
    render: async function* (props, ...args) {
      const element = document.createElement(tag);
      yield element;
      return Promise.all([
        ...Object.keys(attributes ?? {}).map((key) => {
          element.setAttribute(key, '');
          return render(
            attributes[key],
            {
              ...props,
              element,
              key,
              document,
              renderers: ATTRIBUTE_RENDERERS,
            },
            ...args
          );
        }),
        ...children.map((child) => {
          const placeholder = document.createComment(SPECIAL_PLACEHOLDER_ITEM);
          element.appendChild(placeholder);
          return render(
            child,
            {
              ...props,
              placeholder,
              document,
              renderers: ELEMENT_RENDERERS,
            },
            ...args
          );
        }),
      ]);
    },
  });
