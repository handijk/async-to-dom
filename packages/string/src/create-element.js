import { encode as defaultEncode } from 'html-entities';
import { render } from '@async-to-html/render/render.js';
import { safeHtml } from '@async-to-html/render/safe-html/safe-html.js';
import { STRING_RENDERERS } from './renderers/get-default-renderers.js';

export const createElement =
  ({ encode = defaultEncode } = {}) =>
  (tag, attributes, ...children) => ({
    render: async (props, ...args) => {
      const renderedProps = await Promise.all(
        Object.keys(attributes ?? {}).map(
          async (key) =>
            `${key}${`="${await render(
              attributes[key],
              {
                ...props,
                safe: false,
                renderers: STRING_RENDERERS,
                encode,
              },
              ...args
            )}"`}`
        )
      );
      const renderedChildren = await Promise.all(
        children.map((child) =>
          render(
            child,
            {
              ...props,
              safe: true,
              renderers: STRING_RENDERERS,
              encode,
            },
            ...args
          )
        )
      );
      return safeHtml(
        `<${tag} ${renderedProps.join(' ')}>${renderedChildren.join(
          ''
        )}</${tag}>`
      );
    },
  });
