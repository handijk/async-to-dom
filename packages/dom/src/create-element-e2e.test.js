import { describe, expect, test } from 'vitest';
import { createElement as createElementFactory } from './create-element.js';

describe('createElement', () => {
  test('createElement can be aborted', async () => {
    const controller = new AbortController();
    const x = Symbol();
    const args = [{ signal: controller.signal }, x];
    const createElement = createElementFactory({
      document,
    });
    const elementIterator = createElement('div', {}, [
      createElement(
        'span',
        {},
        new Promise((resolve) => setTimeout(() => resolve('to late'), 10000)),
        createElement(
          'span',
          {},
          new Promise((resolve) => setTimeout(() => resolve('on time'), 500))
        )
      ),
    ]).render(...args);
    const { value } = await elementIterator.next();
    document.body.appendChild(value);
    setTimeout(() => {
      controller.abort();
    }, 1000);
    await elementIterator.next();
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
