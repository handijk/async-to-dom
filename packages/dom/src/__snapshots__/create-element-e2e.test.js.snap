// Vitest Snapshot v1

exports[`createElement > createElement can be aborted 1`] = `"<div><span><span>on time</span></span></div>"`;
